core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/InputStreamWithExecutorAsyncRequestBody.java

```diff
@@ -0,0 +1,126 @@
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.reactivestreams.Subscriber;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.annotations.SdkTestInternalApi;
+import software.amazon.awssdk.core.async.AsyncRequestBody;
+import software.amazon.awssdk.core.async.BlockingInputStreamAsyncRequestBody;
+import software.amazon.awssdk.core.internal.util.NoopSubscription;
+import software.amazon.awssdk.utils.IoUtils;
+import software.amazon.awssdk.utils.Logger;
+
+/**
+ * A {@link AsyncRequestBody} that allows reading data off of an {@link InputStream} using a background
+ * {@link ExecutorService}.
+ * <p>
+ * Created via {@link AsyncRequestBody#fromInputStream(InputStream, Long, ExecutorService)}.
+ */
+@SdkInternalApi
+public class InputStreamWithExecutorAsyncRequestBody implements AsyncRequestBody {
+    private static final Logger log = Logger.loggerFor(InputStreamWithExecutorAsyncRequestBody.class);
+
+    private final Object subscribeLock = new Object();
+    private final InputStream inputStream;
+    private final Long contentLength;
+    private final ExecutorService executor;
+
+    private Future<?> writeFuture;
+
+    public InputStreamWithExecutorAsyncRequestBody(InputStream inputStream,
+                                                   Long contentLength,
+                                                   ExecutorService executor) {
+        this.inputStream = inputStream;
+        this.contentLength = contentLength;
+        this.executor = executor;
+        IoUtils.markStreamWithMaxReadLimit(inputStream);
+    }
+
+    @Override
+    public Optional<Long> contentLength() {
+        return Optional.ofNullable(contentLength);
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super ByteBuffer> s) {
+        // Each subscribe cancels the previous subscribe.
+        synchronized (subscribeLock) {
+            try {
+                if (writeFuture != null) {
+                    writeFuture.cancel(true);
+                    waitForCancellation(writeFuture); // Wait for the cancellation
+                    tryReset(inputStream);
+                }
+
+                BlockingInputStreamAsyncRequestBody delegate = AsyncRequestBody.forBlockingInputStream(contentLength);
+                writeFuture = executor.submit(() -> doBlockingWrite(delegate));
+                delegate.subscribe(s);
+            } catch (Throwable t) {
+                s.onSubscribe(new NoopSubscription(s));
+                s.onError(t);
+            }
+        }
+    }
+
+    private void tryReset(InputStream inputStream) {
+        try {
+            inputStream.reset();
+        } catch (IOException e) {
+            throw new UncheckedIOException("Request cannot be retried, because the request stream could not be reset.", e);
```

Comment:
The contract of AsyncRequestBody is related to handling during retries, so it's appropriate to talk about retrying in the error message. As for the exception type, I suppose we could use NonRetryableException to indicate that we shouldn't try again. 

I was trying to be safe here, since I didn't know if the reset failed for a retryable or non-retryable reason, but it's probably not retryable so it would probably be okay to short-circuit future attempts here.